File: backend/src/main/java/com/example/backend/domain/repository/CityRepository.java
package com.example.backend.domain.repository;

import com.example.backend.common.*;
import com.example.backend.domain.City;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;

import java.util.List;

@AppRepository
public interface CityRepository extends JpaRepository<City,Long> {

    @Query(value = """
        SELECT *
        FROM cities c
        WHERE (:region IS NULL OR c.region = :region)
        AND c.population >= :habitantMin
        AND ST_DWithin(
            ST_SetSRID(ST_MakePoint(c.longitude, c.latitude), 4326)::geography,
            ST_SetSRID(ST_MakePoint(:longitude, :latitude), 4326)::geography,
            :rayon
        )
        ORDER BY ST_Distance(
            ST_SetSRID(ST_MakePoint(c.longitude, c.latitude), 4326)::geography,
            ST_SetSRID(ST_MakePoint(:longitude, :latitude), 4326)::geography
        )
    """, nativeQuery = true)
    List<City> getCities(String region, Pageable pageable, int habitantMin, double latitude, double longitude, double rayon);

    @Query("""
        SELECT DISTINCT c.region
        FROM City c
        ORDER BY c.region
    """)
    List<String> getRegion();
}

File: frontend/lib/repository/city_repository.dart
import 'package:frontend/api_service.dart';
import 'package:frontend/models/city.dart';

class CityRepository {
  final ApiService api;

  CityRepository(this.api);

  Future<List<String>> getRegions() async {
    final response = await api.get("/cities/region");
    return List<String>.from(response["data"]);
  }

  Future<List<City>> getCities({
    String? region,
    int? nbVille,
    int? habitantMin,
    double? latitude,
    double? longitude,
    double? rayon,
  }) async {
    final queryParameters = <String, dynamic>{};

    if (region != null && region.isNotEmpty) {
      queryParameters["region"] = region;
    }
    if (nbVille != null && nbVille > 0) {
      queryParameters["nbVille"] = nbVille;
    }
    if (habitantMin != null && habitantMin >= 0) {
      queryParameters["habitantMin"] = habitantMin;
    }
    if (latitude != null) {
      queryParameters["latitude"] = latitude;
    }
    if (longitude != null) {
      queryParameters["longitude"] = longitude;
    }
    if (rayon != null && rayon > 0) {
      queryParameters["rayon"] = rayon;
    }

    final response = await api.get(
      "/cities",
      queryParameters: queryParameters,
    );

    final List<dynamic> data = response["data"];

    return data
        .map((json) => City.fromJson(Map<String, dynamic>.from(json)))
        .toList();
  }
}